File: src/npmctl/debug.py
from __future__ import annotations

import json
from typing import Any

import typer

_DEBUG = False
SENSITIVE_KEYS = {
    "authorization",
    "token",
    "secret",
    "password",
    "dns_provider_credentials",
    "cloudflare_api_token",
}


def set_debug(enabled: bool) -> None:
    global _DEBUG
    _DEBUG = enabled


def is_debug() -> bool:
    return _DEBUG


def log(message: str) -> None:
    if _DEBUG:
        typer.secho(f"[debug] {message}", fg=typer.colors.BRIGHT_BLACK, err=True)


def mask_secret(value: str) -> str:
    if not value or len(value) <= 8:
        return "***"
    return f"{value[:4]}...{value[-4:]}"


def _is_sensitive_key(key_hint: str) -> bool:
    hint = key_hint.lower()
    return any(secret_key in hint for secret_key in SENSITIVE_KEYS)


def redact_value(value: Any, key_hint: str = "") -> Any:
    if _is_sensitive_key(key_hint):
        if isinstance(value, str):
            return mask_secret(value)
        return "***"

    if isinstance(value, dict):
        return {k: redact_value(v, k) for k, v in value.items()}

    if isinstance(value, list):
        return [redact_value(item, key_hint) for item in value]

    if isinstance(value, tuple):
        return tuple(redact_value(item, key_hint) for item in value)

    return value


def preview(value: Any, limit: int = 1600) -> str:
    text = _to_text(value)
    if len(text) <= limit:
        return text
    hidden = len(text) - limit
    return f"{text[:limit]}... [truncated {hidden} chars]"


def _to_text(value: Any) -> str:
    if isinstance(value, (dict, list, tuple)):
        return json.dumps(value, indent=2, ensure_ascii=False)
    return str(value)

File: internal/npmctl/debug.go
package npmctl

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

var sensitiveKeys = []string{
	"authorization",
	"token",
	"secret",
	"password",
	"dns_provider_credentials",
	"cloudflare_api_token",
	"x-api-key",
	"api_key",
	"unifi_api_key",
}

type debugLogger struct {
	enabled bool
	err     io.Writer
}

func (l debugLogger) Log(format string, args ...any) {
	if !l.enabled {
		return
	}
	fmt.Fprintf(l.err, "[debug] "+format+"\n", args...)
}

func maskSecret(value string) string {
	if len(value) <= 8 {
		return "***"
	}
	return value[:4] + "..." + value[len(value)-4:]
}

func isSensitiveKey(key string) bool {
	hint := strings.ToLower(key)
	for _, sensitive := range sensitiveKeys {
		if strings.Contains(hint, sensitive) {
			return true
		}
	}
	return false
}

func redactValue(value any, keyHint string) any {
	if isSensitiveKey(keyHint) {
		if text, ok := value.(string); ok {
			return maskSecret(text)
		}
		return "***"
	}

	switch typed := value.(type) {
	case map[string]any:
		redacted := make(map[string]any, len(typed))
		for key, item := range typed {
			redacted[key] = redactValue(item, key)
		}
		return redacted
	case []any:
		redacted := make([]any, len(typed))
		for i, item := range typed {
			redacted[i] = redactValue(item, keyHint)
		}
		return redacted
	default:
		return value
	}
}

func preview(value any, limit int) string {
	text := toText(value)
	if len(text) <= limit {
		return text
	}
	return fmt.Sprintf("%s... [truncated %d chars]", text[:limit], len(text)-limit)
}

func toText(value any) string {
	switch value.(type) {
	case map[string]any, []any:
		raw, err := json.MarshalIndent(value, "", "  ")
		if err == nil {
			return string(raw)
		}
	}
	return fmt.Sprint(value)
}
